File: wrench-app/src/main/java/com/izettle/wrench/applicationlist/ApplicationViewModel.java
package com.izettle.wrench.applicationlist;


import com.izettle.wrench.database.WrenchApplication;
import com.izettle.wrench.database.WrenchApplicationDao;

import javax.inject.Inject;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

public class ApplicationViewModel extends ViewModel {

    private final MediatorLiveData<PagedList<WrenchApplication>> mediatedApplications;

    private final MutableLiveData<Boolean> listEmpty;

    @Inject
    ApplicationViewModel(WrenchApplicationDao applicationDao) {

        listEmpty = new MutableLiveData<>();
        listEmpty.setValue(true);

        LiveData<PagedList<WrenchApplication>> applications =
                new LivePagedListBuilder<>(applicationDao.getApplications(),
                        new PagedList.Config.Builder()
                                .setEnablePlaceholders(true)
                                .setPageSize(10)
                                .setPrefetchDistance(10)
                                .build()).build();

        mediatedApplications = new MediatorLiveData<>();

        mediatedApplications.addSource(applications, wrenchApplications -> {
            listEmpty.setValue(wrenchApplications == null || wrenchApplications.size() == 0);
            mediatedApplications.setValue(wrenchApplications);
        });
    }

    LiveData<PagedList<WrenchApplication>> getApplications() {
        return mediatedApplications;
    }

    LiveData<Boolean> isListEmpty() {
        return listEmpty;
    }
}

File: wrench-app/src/main/java/com/izettle/wrench/applicationlist/ApplicationViewModel.kt
package com.izettle.wrench.applicationlist


import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import com.izettle.wrench.database.WrenchApplication
import com.izettle.wrench.database.WrenchApplicationDao

class ApplicationViewModel
internal constructor(applicationDao: WrenchApplicationDao) : ViewModel() {

    private val mediatedApplications: MediatorLiveData<PagedList<WrenchApplication>>

    private val listEmpty: MutableLiveData<Boolean> = MutableLiveData()

    internal val applications: LiveData<PagedList<WrenchApplication>>
        get() = mediatedApplications

    internal val isListEmpty: LiveData<Boolean>
        get() = listEmpty

    init {

        listEmpty.value = true

        val applications = LivePagedListBuilder(applicationDao.getApplications(),
                PagedList.Config.Builder()
                        .setEnablePlaceholders(true)
                        .setPageSize(10)
                        .setPrefetchDistance(10)
                        .build()).build()

        mediatedApplications = MediatorLiveData()

        mediatedApplications.addSource(applications) { wrenchApplications ->
            listEmpty.value = wrenchApplications == null || wrenchApplications.size == 0
            mediatedApplications.setValue(wrenchApplications)
        }
    }
}
